### PR TITLE
Handle return for numbered lists

### PR DESCRIPTION
**Pull request by lukas-reineke**
I started working on this before you pushed your version

But my version also updates the numbers of all following items, when you add one in the middle.

To get this to work nicer, I also had to make a change to the treesitter grammar.
That PR is open here: https://github.com/milisims/tree-sitter-org/pull/5

As a bonus, this also handles multi line list items.

### CODE REVIEW

File: lua/orgmode/org/mappings.lua
Comment (kristijanhusak) [inline lua/orgmode/org/mappings.lua:254]: Can this ever not be `listitem`? Does it make sense to break out if this is false?

**Review by kristijanhusak — COMMENTED** [review]
I'll wait for the upstream PR to be merged first before testing this out.

File: lua/orgmode/org/mappings.lua
Comment (lukas-reineke) [reply lua/orgmode/org/mappings.lua:254]: `next_sibling` returns all nodes, named and unamed. So it can be the newline at the end of each list item.
I'll change it to `ts_utils.get_next_node`, that only returns named nodes, that is easier to understand.

**Comment by lukas-reineke** [comment]
I added a bit more stuff.

All lists and headlines now support multiline. Doesn't matter what is nested in them. On enter they will create a new item on the same level they are, after the end of their section.

File: lua/orgmode/org/mappings.lua
Comment (kristijanhusak) [inline lua/orgmode/org/mappings.lua:221]: Why is this changed? `stars`, `item` and `tags` nodes are all on the same level, so this should handle return from all of those.
Also, below `section` variable will not return section node if you are on `item` or `tags` node, since this will go one parent up, and that will go 2 more parents up, which can end up on the top level document. Maybe I'm missing something, didn't tested it yet.

File: lua/orgmode/org/mappings.lua
Comment (kristijanhusak) [inline lua/orgmode/org/mappings.lua:233]: I think this should be `normal! ^`, since `normal` uses mapped key without `bang`. Also, does `startofline` option affect this functionality?

File: lua/orgmode/utils.lua
Comment (kristijanhusak) [inline lua/orgmode/utils.lua:264]: When is node empty string? This function should either get a ts node or nil.

**Review by kristijanhusak — DISMISSED** [review]

File: lua/orgmode/org/mappings.lua
Comment (lukas-reineke) [reply lua/orgmode/org/mappings.lua:221]: The logic before was a bit broken, but still worked.
If it was not a `headline` type, it does `ts_utils.goto_node(item.node:parent())`, which goes to the start of the headline. But then `Files.get_current_file():get_current_node()` returns the smallest node at that point, which is `stars`.
But that's fine, as stars also has the level.
But now for the change I made, I need to know which level I am on, so I can go up to the section. So if the cursor is between the stars and the line, it is type `headline`. In that case I do the same as before to change `item` to `stars`.

I agree, it is not really obvious. Maybe the code should just walk up the parent until it reaches `section` to be easier to understand.

File: lua/orgmode/org/mappings.lua
Comment (lukas-reineke) [reply lua/orgmode/org/mappings.lua:233]: good point about `!`
I think `startofline` should not affect this.

File: lua/orgmode/utils.lua
Comment (lukas-reineke) [reply lua/orgmode/utils.lua:264]: I don't really know why this happened. This is unrelated to the rest of my changes.
It had a state where this throws errors, even on the current `tree-sitter` branch.

I'll see if I can reproduce this, and see where it comes from

File: lua/orgmode/org/mappings.lua
Comment (kristijanhusak) [reply lua/orgmode/org/mappings.lua:221]: > I agree, it is not really obvious. Maybe the code should just walk up the parent until it reaches section to be easier to understand.


I like this. You can introduce a helper in utils that would do this traversing to closest parent, something like good ol' Jquery's [parentsUntil](https://api.jquery.com/parentsuntil/)

File: lua/orgmode/utils.lua
Comment (kristijanhusak) [reply lua/orgmode/utils.lua:264]: Yeah, please try to reproduce it. If it's not related to this PR, i'll fix it separately.

Comment (lukas-reineke) [reply lua/orgmode/utils.lua:264]: It happens when there is an empty headline.

just opening a file like this throws.

```org
*
```

Comment (kristijanhusak) [reply lua/orgmode/utils.lua:264]: Thanks, I'll take a look.

Comment (kristijanhusak) [reply lua/orgmode/utils.lua:264]: Fixed it. I was using default as empty string instead of nil for todo keyword node, which is missing in this case.

**Review by kristijanhusak — COMMENTED** [review]
This looks ok for now. Once grammar PR is merged, I'll give this a test and merge it. Thanks!

**Comment by kristijanhusak** [comment]
Since upstream PR is now merged, I'll start testing this manually tomorrow. 
Today I added some UI tests that does testing of the mappings. I would appreciate if you would rebase and add a test for this, so we have it covered. Here's the current test that tests my simple version of numbered list increase: https://github.com/kristijanhusak/orgmode.nvim/blob/tree-sitter/tests/plenary/ui/mappings_spec.lua#L330. Feel free to add any other tests if you think is needed.

**Comment by lukas-reineke** [comment]
oh `next_named_sibling` is Neovim new in 0.6.
I'll revert this back to use the `ts_util` helper

**Comment by lukas-reineke** [comment]
I added some tests, let me know if there is anything else you'd like tested.

Also about the tests, I think it would be better to have dedicated fixtures for tests that jump around the whole file.
It is inconvenient to have to update unrelated tests when you want to add a new test case.

**Comment by kristijanhusak** [comment]
> I added some tests, let me know if there is anything else you'd like tested.
> 
> Also about the tests, I think it would be better to have dedicated fixtures for tests that jump around the whole file.
> It is inconvenient to have to update unrelated tests when you want to add a new test case.

Thanks! Looks great!

Yeah, I noticed that issue with tests. I'll switch to creating a fixture for each separate test. It will be slower, but much less fragile and unstable.

Once I do some manual tests around this I'll merge it (hopefully today).

**Comment by kristijanhusak** [comment]
I gave this a test, works great!
Thank you for your contribution!

**Pull request merged**
